diff -upr lsof_4.82-rh.old/dialects/linux/dproc.c lsof_4.82-rh/dialects/linux/dproc.c
--- lsof_4.82-rh.old/dialects/linux/dproc.c	2012-02-22 16:33:41.000000000 +0100
+++ lsof_4.82-rh/dialects/linux/dproc.c	2012-02-22 16:35:17.298691514 +0100
@@ -1341,7 +1341,7 @@ process_proc_map(p, s, ss)
 		mss = SB_DEV | SB_INO | SB_MODE;
 		if (ds)
 		    alloc_lfile("DEL", -1);
-		else if (!efs) {
+		else if (!efs && !Fwarn) {
 		    (void) snpf(nmabuf, sizeof(nmabuf), "(stat: %s)",
 			strerror(en));
 		    nmabuf[sizeof(nmabuf) - 1] = '\0';
