diff -up lsof_4.82-rh/arg.c.multi-e-option lsof_4.82-rh/arg.c
--- lsof_4.82-rh/arg.c.multi-e-option	2011-10-31 17:38:15.000000000 +0100
+++ lsof_4.82-rh/arg.c	2011-10-31 17:40:07.451977416 +0100
@@ -778,8 +778,8 @@ enter_efsys(e, rdlnk)
 	ep->pathl = i;
 	ep->rdlnk = rdlnk;
 	ep->mp = (struct mounts *)NULL;
-	if (!(ep->next = Efsysl))
-	    Efsysl = ep;
+	ep->next = Efsysl;
+	Efsysl = ep;
 	return(0);
 }
 #endif	/* defined(HASEOPT) */
